### DISCUSSION TITLE
Dependency Dashboard improvements

### DISCUSSION THREAD

**Discussion by rcdailey** [renovatebot/renovate#29284] [Suggest an Idea]
I love the dependency dashboard, but I think it lacks a few things:

- When there is a list of PRs pending creation due to status checks not passing, provide a list of all of those status checks that are pending.
- Provide a link to the project's dashboard on `developer.mend.io`.
- Include a timestamp in the description for when the last scan took place on the repository.

Thanks for a fabulous project. It has completely changed the way I manage dependencies in multiple projects of mine!

**Comment by HonkingGoose** [comment]
## List pending status checks on Dependency Dashboard

> - When there is a list of PRs pending creation due to status checks not passing, provide a list of all of those status checks that are pending.

Fetching all the statuses may spend too much compute time or GitHub API calls. We also need to stay within certain character limits for the text in the issue body, or be smart about what to show when we reach the limit.

### Maybe link to test overview on GitHub?

GitHub has a page that has a overview of all the tests that ran on a branch. Maybe we can add that link to the Dependency Dashboard, like this:

[Summary of tests for the `ghcr.io-renovatebot-base-image-6.x` branch](https://github.com/renovatebot/renovate/actions/runs/9256203577).

I don't know if we can tell Renovate where/how to find the correct overview page. And maybe fetching the correct overview page is too expensive as well. :upside_down_face: 

I'll let the maintainers respond officially.

## More context about character limits

- #17749
- #18244
- #22156
- #16373
- #16527
- #15850
- #16959

**Reply by rcdailey** [reply]
I mean, I think it would be reasonable to offer some options to help subtract information from the description, too. For example, the "Detected Dependencies" list is the same constantly and I don't find it useful personally so an option to disable that might help free up some character limit. Likewise, an option to enable the status checks also seems reasonable. Let people mix & match what information they'd like to see. There's obviously not a one-size-fits-all dashboard; so customization/configuration seems like the logical solution.

**Comment by rarkins** [comment]
> * When there is a list of PRs pending creation due to status checks not passing, provide a list of all of those status checks that are pending.

If these are Renovate's checks, this might be possible. But it's not something I think is scalable/reasonable if you mean for us to list every one of your custom CI checks, especially because that list might be out of date within minutes and people will think it's a bug.

> * Provide a link to the project's dashboard on `developer.mend.io`.

This can/should be done.

> * Include a timestamp in the description for when the last scan took place on the repository.

I prefer to not have this because it would mean we need to edit the Dashboard every single run, even if nothing has changed. Possibly embedding a "badge" pointing to the mend platform which could include recent jobs or something like that would be feasible.

**Reply by rcdailey** [reply]
> If these are Renovate's checks, this might be possible. But it's not something I think is scalable/reasonable if you mean for us to list every one of your custom CI checks, especially because that list might be out of date within minutes and people will think it's a bug.

This is a reasonable concern, and I'll admit I don't know how far down the rabbit hole goes on this. I was thinking of a very specific example for my use case, which is that I have configured `"minimumReleaseAge": "5 days"` and it took me a while to understand what exactly it was waiting on.

Instead of proposing a _solution_, I'll just point out the _problem_: It's currently unintuitive to look at the list of pending PRs and understand what is holding up those from being created. It could be anything. As far as I know, there's not even a way to look at the mend dashboard to see what is causing it. Too many times I felt I've had to dig into the Renovate logs to find information and I am 100% of the strong opinion that that's a terrible user experience (intentional or not).

I'd like a solution that helps make it more obvious what status checks are pending. If that requires me to click a link to a new "Pending Status Checks" tab on the mend dashboard, I'd be fine with that. Software engineering is all about coming up with creative solutions. I am not fixated on a specific solution I just want it to be more intuitive.

> I prefer to not have this because it would mean we need to edit the Dashboard every single run, even if nothing has changed. Possibly embedding a "badge" pointing to the mend platform which could include recent jobs or something like that would be feasible.

Again, valid concern, and I'm learning more as you point out these issues. Similar to the first point, I probably should have focused on the problem rather than a solution.

The problem in this case is that I don't have a way, without visiting the Mend dashboard, to see how "out of date" the dashboard is. I suppose I could click the "Edited" drop down on the issue to see the change history, but this doesn't necessarily represent the last run I suppose. My solution focused on editing the dashboard because 1) I assumed that wouldn't be an issue and 2) I like the idea of centralizing information and avoiding clicks/redirections when possible.

I think as long as the problem statements are agreeable, solutions can be left up to you folks. You've done a fantastic job with Renovate so far and I have no doubt whatever solution you come up with will be satisfactory (to me at least!).

Thanks for reading my feedback, seriously considering it, and responding so quickly!

**Reply by HonkingGoose** [reply]
> > * Provide a link to the project's dashboard on `developer.mend.io`.
> 
> This can/should be done.

See issue:

-  #29314

**Comment by Thenlie** [comment]
I would also love to see `current version -> upgrade version` as is done in the PR's. It can be unclear on the dashboard how big of an upgrade each PR is. 

Honestly, I would just prefer to see the entire table from the PR in the dashboard. This gives us links to the diff, link to the source repo, current and upgrade version, as well as some other useful info. Photo attached of the table I am referring to.
![image](https://github.com/renovatebot/renovate/assets/41388783/5aeaadf6-2126-4fcb-bbbc-7e97deb4a8a9)

**Reply by HonkingGoose** [reply]
## Show current version and upgrade version on Dependency Dashboard

> I would also love to see `current version -> upgrade version` as is done in the PR's. It can be unclear on the dashboard how big of an upgrade each PR is.

This sounds doable. But as usual, I'll let a maintainer respond officially, and let them give the go/no go signal. :wink: 

### Proposed pattern

How about a pattern like this?

```markdown
- [ ] `semanticCommitType` + `depType`: update `depNameLinked` from `currentValue` to `newValue`
```

#### Mockup of pattern

Here's a real-world example:

- [ ] chore(deps): update pnpm from `9.3.0` to `v9.4.0` 

```markdown
- [ ] chore(deps): update pnpm from `9.3.0` to `v9.4.0` 
````

If Renovate does not know the old version (or Renovate can't "see" the old update), the new version is still always at the _end_ of the line. That way, the last thing you read is the key thing to remember: the newest version number. :smile: 

## Copy/pasting the table from the PR into the Dashboard is difficult

> Honestly, I would just prefer to see the entire table from the PR in the dashboard. This gives us links to the diff, link to the source repo, current and upgrade version, as well as some other useful info. Photo attached of the table I am referring to.
> ![Renovate PR body, table with information about a update](https://private-user-images.githubusercontent.com/41388783/340843935-5aeaadf6-2126-4fcb-bbbc-7e97deb4a8a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4MDM2NjIsIm5iZiI6MTcxODgwMzM2MiwicGF0aCI6Ii80MTM4ODc4My8zNDA4NDM5MzUtNWFlYWFkZjYtMjEyNi00ZmNiLWJiYmMtN2U5N2RlYjRhOGE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDEzMjI0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ2NzRiYmU2Y2U4NDE2NTg5YmZmYzgzYmE1N2E0ODc2MjQwODRjMWZiZWQxZjk2NGU2YmY3NzEwMTNlNjRkYTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Q4RYFtUhFD-1Y8sFYnqPAT5EA7UO0JAX6DFqk_AHd_o)

I get where you're coming from. And I agree that it would be nice to have this info in the Dependency Dashboard. But this is a hard problem to solve.

### My experiences mocking up a better Dashboard

I have mocked up the Dependency Dashboard a few times, it usually ends with deciding that the improvement is too hard to build, or does not work properly on all platforms. Here's my conclusion at the end of a mock up sequence:

- https://github.com/renovatebot/renovate/issues/10924#issuecomment-917377915

### Why we can't show all the things

With the Dependency Dashboard we need:

- [ ] Interactive check-boxes (which do _not_ work properly with table layouts...)
- [ ] To show enough useful information
- [ ] Stay within the Git hosting platforms's character limits (if possible)
- [ ] Be smart about what to drop when we hit the character limits

## Related issue

We do have this open issue about showing more details in the Dependency Dashboard:

- https://github.com/renovatebot/renovate/issues/17749

Maybe my idea of rewording the checkbox text string so it show the "from ... to ..." bit can go in that issue. As usual though, that's for the maintainers to approve. :wink:
